File: my-folder/0207-course-schedule/solution.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        // adj list
        List<Integer>[] graph = new List[numCourses];
        for(int i=0;i<numCourses;i++){
            graph[i] = new ArrayList<>();
        }
        
        int[] indegree = new int[numCourses];
        for(int[] pair: prerequisites){
            // u -> v
            int u = pair[1];
            int v = pair[0];
            indegree[v]++;
            graph[u].add(v);
        }
        
        Queue<Integer> q = new LinkedList<>();
        // nodes with indegree being zero
        for(int i=0;i<numCourses;i++){
            if(indegree[i]==0){
                q.offer(i);
            }
        }
        List<Integer> sorted = new ArrayList<>();
        while(!q.isEmpty()){
            int node = q.poll();
            sorted.add(node);
            for(int next: graph[node]){
                indegree[next]--;
                if(indegree[next]==0){
                    q.offer(next);
                }
            }
            
        }
        return sorted.size()==numCourses;
    }
}

File: my-folder/0207-course-schedule/solution.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        grid = [[] for _ in range(numCourses)]
        for t,s in prerequisites:
            grid[s].append(t)
        self.path = [False for _ in range(numCourses)]
        self.visited = [False for _ in range(numCourses)]
        self.has_cycle = False
        for i in range(numCourses):
            self.traverse(grid, i)
        return not self.has_cycle

    def traverse(self, grid, i):
        if self.path[i]:
            self.has_cycle = True
        if self.visited[i] or self.has_cycle:
            return
        self.path[i] = True
        self.visited[i] = True
        for v in grid[i]:
            self.traverse(grid, v)
        self.path[i] = False

